src/main/java/org/broadinstitute/hellbender/utils/pileup/ReadPileup.java

```diff
@@ -0,0 +1,274 @@
+        final int[] counts = new int[4];
+
+        for (final PileupElement pile : this) {
+            // skip deletion sites
+            if (!pile.isDeletion()) {
+                final int index = BaseUtils.simpleBaseToBaseIndex(pile.getBase());
+                if (index != -1) {
+                    counts[index]++;
+                }
+            }
+        }
+
+        return counts;
+    }
+
+    @Override
+    public String toString() {
+        return String.format("%s %s %s %s",
+                loc.getContig(),
+                loc.getStart(),
+                new String(getBases()),
+                getQualsString());
+    }
+
+    /**
+     * Returns a list of the reads in this pileup. Note this call costs O(n) and allocates fresh lists each time
+     */
+    public List<GATKRead> getReads() {
+        return pileupElements.stream().map(pe -> pe.getRead()).collect(Collectors.toList());
+    }
+
+    /**
+     * Returns the number of elements that satisfy the predicate.
+     */
+    public int getNumberOfElements(final Predicate<PileupElement> peFilter){
+        Utils.nonNull(peFilter);
+        //Note: pileups are small so int is fine.
+        return (int)pileupElements.stream().filter(peFilter).count();
+    }
+
+    /**
+     * Returns a list of the offsets in this pileup.
+     * Note: this call costs O(n) and allocates fresh lists each time
+     */
+    public List<Integer> getOffsets() {
+        return pileupElements.stream().map(pe -> pe.getOffset()).collect(Collectors.toList());
+    }
+
+    //Extracts an int array by mapping each element in the pileup to int.
+    private int[] extractIntArray(final ToIntFunction<PileupElement> map){
+        return pileupElements.stream().mapToInt(map).toArray();
+    }
+
+    /**
+     * Returns an array of the bases in this pileup.
+     * Note: this call costs O(n) and allocates fresh array each time
+     */
+    public byte[] getBases() {
+        return toByteArray(extractIntArray(pe -> pe.getBase()));
+    }
+
+    /**
+     * Returns an array of the quals in this pileup.
+     * Note: this call costs O(n) and allocates fresh array each time
+     */
+    public byte[] getBaseQuals() {
+        return toByteArray(extractIntArray(pe -> pe.getQual()));
+    }
+
+    //Converts array of ints to array of bytes by hard casting (loses precision if ints are large).
+    private byte[] toByteArray(final int[] ints) {
+        final byte[] bytes = new byte[ints.length];
+        for (int i = 0; i < ints.length; i++) {
+            bytes[i] = (byte)ints[i];
+        }
+        return bytes;
+    }
+
+    /**
+     * Get an array of the mapping qualities.
+     */
+    public int[] getMappingQuals() {
+        return extractIntArray(pe -> pe.getMappingQual());
+    }
+
+    private String getQualsString() {
+        final byte[] quals = getBaseQuals();
+        for (int i = 0; i < quals.length; i++) {
+            quals[i] = (byte) (33 + quals[i]);  //as per SAM spec
```

Comment:
Do we understand why the previous version of this code was capping the quals at 63 rather than converting to ASCII as the current version does?